crates/services/consensus_module/src/block_verifier.rs

```diff
@@ -0,0 +1,185 @@
+//! The module provides the functionality that verifies the blocks and headers based
+//! on the used consensus.
+
+pub mod config;
+pub mod ports;
+
+use crate::block_verifier::{
+    config::Config,
+    ports::Database,
+};
+use anyhow::ensure;
+use fuel_core_chain_config::{
+    BlockProduction,
+    PoABlockProduction,
+};
+use fuel_core_types::{
+    blockchain::{
+        block::Block,
+        consensus::Consensus,
+    },
+    fuel_types::Bytes32,
+    tai64::{
+        Tai64,
+        Tai64N,
+    },
+};
+use std::ops::Add;
+
+/// Verifier is responsible for validation of the blocks and headers.
+pub struct Verifier<D, R> {
+    config: Config,
+    database: D,
+    _relayer: R,
+}
+
+impl<D, R> Verifier<D, R> {
+    /// Creates a new instance of the verifier.
+    pub fn new(config: Config, database: D, _relayer: R) -> Self {
+        Self {
+            config,
+            database,
+            _relayer,
+        }
+    }
+}
+
+impl<D, R> Verifier<D, R>
+where
+    D: Database,
+{
+    /// Verifies **all** fields of the block based on used consensus to produce a block.
+    ///
+    /// Return an error if the verification failed, otherwise `Ok(())`.
+    pub async fn verify_block_fields(
+        &self,
+        consensus: &Consensus,
+        block: &Block,
+    ) -> anyhow::Result<()> {
+        match consensus {
+            Consensus::Genesis(_) => self.verify_genesis_block_fields(block),
+            Consensus::PoA(_) => self.verify_poa_block_fields(block),
+        }
+    }
+
+    fn verify_genesis_block_fields(&self, block: &Block) -> anyhow::Result<()> {
+        let expected_genesis_height = self
+            .config
+            .chain_config
+            .initial_state
+            .as_ref()
+            .map(|config| config.height.unwrap_or_else(|| 0u32.into()))
+            .unwrap_or_else(|| 0u32.into());
+        let actual_genesis_height = *block.header().height();
+
+        ensure!(
+            block.header().prev_root() == &Bytes32::zeroed(),
+            "The genesis previous root should be zeroed"
+        );
+        ensure!(
+            block.header().time() == Tai64::UNIX_EPOCH,
+            "The genesis time should be unix epoch time"
+        );
+        ensure!(
+            // TODO: Set `da_height` based on the chain config.
+            block.header().da_height == Default::default(),
+            "The genesis `da_height` is not as expected"
+        );
+        ensure!(
+            expected_genesis_height == actual_genesis_height,
+            "The genesis height is not as expected"
+        );
+        Ok(())
+    }
+
+    fn verify_poa_block_fields(&self, block: &Block) -> anyhow::Result<()> {
```

Comment:
todo: this kind of logic should be owned by the respective CM impl (e.g. done in PoA).